README.md
# numberreader

A Clojure library designed to read a number between 0 and 999999 inclusive, and translate it into a gramatically correct representation.

For example, the library would translate 111 into "one hundred and eleven" and 64 into "sixty-four". The output is given as a string.

If the number is not between 0 and 999999 inclusive, an exception will be thrown.

Documentation can be found in target/doc/ and was generated using [codox](https://github.com/weavejester/codox).

## Usage

Download the library, navigate to the top level, i.e. - the directory containing project.clj and start the REPL.

In the REPL, type (require '[numberreader.core :refer :all]), then type (up-to-mill number) for whatever number you like.

Typical output is shown below:

=> (up-to-mill 523)
"five hundred and twenty-three"

=>(up-to-mill 1000)
"one thousand"

=> (up-to-mill -12)
Exception Your number is not between 0 and 999999.  numberreader.core/number-to-string (core.clj:91)

## License

Copyright © 2016 Alan Forrester

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased][unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2016-04-20
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2016-04-20
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[unreleased]: https://github.com/your-name/numberreader/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/numberreader/compare/0.1.0...0.1.1


test/numberreader/core_test.clj
(ns numberreader.core-test
  (:require [clojure.test :refer :all]
            [numberreader.core :refer :all]))

(deftest thousands-test
  (are [x y] (= x y)
       (thousands 1000) 1
       (thousands 2430) 2
       (thousands 250) 0
       (thousands 35) 0
       (thousands 3) 0))

(deftest hundreds-test
  (are [x y] (= x y)
       (hundreds 5000) 0
       (hundreds 3520) 5
       (hundreds 230) 2
       (hundreds 93) 0
       (hundreds 6) 0))

(deftest tens-test
  (are [x y] (= x y)
       (tens 3400) 0
       (tens 7630) 30
       (tens 305) 5
       (tens 43) 43
       (tens 4) 4))

(deftest tens-to-string-test
  (are [x y] (= x y)
       (tens-to-string 53) "fifty-three"
       (tens-to-string 10) "ten"))

(deftest hundreds-to-string-test
  (are [x y] (= x y)
       (hundreds-to-string 0) ""
       (hundreds-to-string 3) "three hundred"))

(deftest thousands-to-string-test
  (are [x y] (= x y)
       (thousands-to-string 0) ""
       (thousands-to-string 1) "one thousand"
       (thousands-to-string 831) "eight hundred and thirty-one thousand"))

(deftest tens-hundreds-to-string-test
  (are [x y] (= x y)
       (tens-hundreds-to-string 341) "three hundred and forty-one"
       (tens-hundreds-to-string 300) "three hundred"
       (tens-hundreds-to-string 0) "zero"))

(deftest combine-lower-higher-test
  (are [x y] (= x y)
       (combine-lower-higher "three hundred"  "ten" " and ") "three hundred and ten"
       (combine-lower-higher "seven hundred"  "zero" " and ") "seven hundred"
       (combine-lower-higher "one thousand" "six hundred" " ") "one thousand six hundred"))

(deftest out-of-range-test
  (are [x y] (= x y)
       (out-of-range 10 0 1000) false
       (out-of-range 5 7 34) true
       (out-of-range 345 62 250) true)
  (is (thrown? Exception #"Your lower limit is larger than your upper limit."
               (out-of-range 1 5 2))))

(deftest up-to-mill-test
    (are [x y] (= x y)
       (up-to-mill 523) "five hundred and twenty-three"
       (up-to-mill 0) "zero"
       (up-to-mill 1000) "one thousand"
       (up-to-mill 12001) "twelve thousand and one"
       (up-to-mill 12041) "twelve thousand and forty-one"
       (up-to-mill 173345) "one hundred and seventy-three thousand three hundred and forty-five")
  (is (thrown? Exception #"Your number is not between 0 and 999999."
               (number-to-string -12)))
  (is (thrown? Exception #"Your number is not between 0 and 999999."
               (number-to-string 153000000))))


src/numberreader/core.clj
(ns numberreader.core)

(def numbers-dictionary
  "A hash-map that contains numbers as keys and string representations of those numbers as values."
  {0 "zero"
   1 "one"
   2 "two"
   3 "three"
   4 "four"
   5 "five"
   6 "six"
   7 "seven"
   8 "eight"
   9 "nine"
   10 "ten"
   11 "eleven"
   12 "twelve"
   13 "thirteen"
   14 "fourteen"
   15 "fifteen"
   16 "sixteen"
   17 "seventeen"
   18 "eighteen"
   19 "nineteen"
   20 "twenty"
   30 "thirty"
   40 "forty"
   50 "fifty"
   60 "sixty"
   70 "seventy"
   80 "eighty"
   90 "ninety"})

(defn magnitude-coefficient
  "Gives factor in front of terms of two given magnitudes.
  For example, (magnitude-coefficient 340594 100000 1000) would give 340."
  [number upper lower]
  (let [left (rem number upper)]
    (int (/ left lower))))

(defn thousands
  "Gives the number of 1000s in number."
  [number]
  (magnitude-coefficient number 1000000 1000))

(defn hundreds
  "Gives the number of 100s in number."
  [number]
  (magnitude-coefficient number 1000 100))

(defn tens
  "Gives what is left over from number after the thousands and hundreds have been removed."
  [number]
  (magnitude-coefficient number 100 1))

(defn tens-to-string
  "Takes a number in the range 0 to 99 and produces a string representing that number."
  [tes]
  (if (numbers-dictionary tes)
    (numbers-dictionary tes)
    (let
      [ones (numbers-dictionary (rem tes 10))
       ts (-> tes
              (#(/ % 10))
              (int)
              (#(* % 10))
              (numbers-dictionary))]
      (str ts "-" ones))))

(defn mag-to-string
  "Takes a number.
  Returns a blank string if zero and and the string (str mag st) otherwise."
  [mag st f]
  (if (zero? mag)
    ""
    (str (f mag) st)))

(defn combine-lower-higher
  "Takes strings representing lower and higher numbers and combines them using a connector.
  If higher is \"one hundred\", lower is \"fifty-three\" and the connector is \" and \",
  then we get \"one hundred and fifty-three\"."
  [higher lower connector]
  (cond
    (empty? higher) lower
    (= "zero" lower) higher
    :else (str higher connector lower)))

(defn hundreds-to-string
  "Takes a number in the range 0 to 9.
  Returns a blank string if zero and the string \"hs hundred\" otherwise."
  [hs]
  (mag-to-string hs " hundred" numbers-dictionary))

(defn out-of-range
  "Checks whether the number is greater than lower and less than upper.
  Throws an exception if the lower is larger than upper."
  [number lower upper]
  (if (> lower upper)
    (throw (Exception. "Your lower limit is larger than your upper limit."))
    (or (> number upper) (< number lower))))

(defn join-mags
  [s1 s2]
    (combine-lower-higher s1 s2 " "))

(defn number-to-string
  "Converts a number to a gramatically correct string if it is between 0 and maximum,
  given functions for mapping the various orders of magnitude to strings
  and a function for joining the strings."
  [number maximum numstringfs joinfunc]
  (if (out-of-range number 0 maximum)
    (throw (Exception. (str "Your number is not between 0 and " maximum ".")))
    (let [mags ((apply juxt numstringfs) number)
          most (reduce joinfunc "" (butlast mags))]
      (combine-lower-higher (clojure.string/trim most) (last mags) " and "))))

(defn tens-hundreds-to-string
  "Works out strings for tens and hundreds and then joins them."
  [number]
  (number-to-string
    number
    999
    [(comp hundreds-to-string hundreds)
     (comp tens-to-string tens)]
    join-mags))

(defn thousands-to-string
  "Takes a number in the range 1 to 999 converts it with tens-hundreds-to-string
  and adds the string \" thousand\" to the end."
  [ths]
  (mag-to-string ths " thousand" tens-hundreds-to-string))

(defn up-to-mill
  "Converts a number to a gramatically correct string if it is between 0 and 1 million-1 inclusive."
  [number]
  (number-to-string
    number
    999999
    [(comp thousands-to-string thousands)
     (comp hundreds-to-string hundreds)
     (comp tens-to-string tens)]
    join-mags))
